cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java

```diff
@@ -367,10 +373,65 @@ public void testDelete() throws Exception {
                                          TEST_NAMESPACE1));
     Assert.assertEquals(200, response.getStatusLine().getStatusCode());
 
-    List<ArtifactSummary> summaries = readResponse(response, new TypeToken<List<ArtifactSummary>>() { }.getType());
+    List<ArtifactSummary> summaries = readResponse(response, new TypeToken<List<ArtifactSummary>>() {
+    }.getType());
     Assert.assertFalse(summaries.isEmpty());
   }
 
+  /**
+   * Tests deleting an application.
+   */
+  @Test
+  public void testDeleteVersionedApp() throws Exception {
+    // Delete an non-existing app with version
+    HttpResponse response = doDelete(getVersionedAPIPath("apps/XYZ/versions/" + ApplicationId.DEFAULT_VERSION,
+                                                         Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE1));
+    Assert.assertEquals(404, response.getStatusLine().getStatusCode());
+
+    // Start a fow for the App
+    deploy(WordCountApp.class, Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE1);
+    ApplicationId appId = new ApplicationId(TEST_NAMESPACE1, "WordCountApp");
+    ProgramId program = appId.program(ProgramType.FLOW, "WordCountFlow");
+    startProgramVersioned(program, ImmutableMap.<String, String>of(), 200);
+    waitStateVersioned(program, "RUNNING");
+    // Try to delete an App while its flow is running
+    response = doDelete(getVersionedAPIPath(
+      String.format("apps/%s/versions/%s", appId.getApplication(), appId.getVersion()),
+      Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE1));
+    Assert.assertEquals(409, response.getStatusLine().getStatusCode());
+    Assert.assertEquals("'" + program.getParent() + "' could not be deleted. Reason: The following programs are" +
+                          " still running: "
+                          + program.getProgram(), readResponse(response));
+
+    stopProgramVersioned(program, null, 200, null);
+    waitStateVersioned(program, "STOPPED");
+
+    //Delete the App after stopping the flow
+    response = doDelete(getVersionedAPIPath(
+      String.format("apps/%s/versions/%s", appId.getApplication(), appId.getVersion()),
+      Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE1));
+    Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+    response = doDelete(getVersionedAPIPath(
+      String.format("apps/%s/versions/%s", appId.getApplication(), appId.getVersion()),
+      Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE1));
+    Assert.assertEquals(404, response.getStatusLine().getStatusCode());
+
+    // Delete the app in the wrong namespace
+    response = doDelete(getVersionedAPIPath(
+      String.format("apps/%s/versions/%s", appId.getApplication(), appId.getVersion()),
+      Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE2));
+    Assert.assertEquals(404, response.getStatusLine().getStatusCode());
+
+    // TODO: deleting the app should not delete the artifact
```

Comment:
Add a JIRA for adding more tests here. I think we need more tests like deploying/starting/stopping/deleting multiple versions of same app (including non-versioned app).